### PR TITLE
add new dedup strategy

### PR DESCRIPTION
**Pull request by yaojiefeng**
Add a strategy to preserve records from the scd2 target table.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 1]
```diff
@@ -89,6 +95,69 @@ protected StructuredRecord computeNext() {
       }
     }
 
+    // get the next record that is available to compute the end time, it must be a record with same id
```
Comment (ghost) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: This method becomes over 100 lines. Can we structure it into smaller ones?

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 2]
```diff
@@ -58,11 +59,16 @@ protected StructuredRecord computeNext() {
 
     prev = cur;
     cur = next != null ? next : records.next();
-    next = records.hasNext() ? records.next() : null;
+    next = nextNext != null ? nextNext : (records.hasNext() ? records.next() : null);
```
Comment (ghost) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: Should we reset nextNext here? It's possible to exit the current method without recomputing nextNext. Not sure if this would affect algorithm as next would be equal to nextNext on exit and it does not seem like a right state.

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 1 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: checkstyle for these projects is 120 characters a line

File: src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java [hunk 3]
```diff
@@ -277,6 +298,30 @@ private void validate(@Nullable Schema actualSchema, FailureCollector failureCol
...
+
+        if (isTargetField != null) {
+          Schema.Field field = actualSchema.getField(isTargetField);
+          if (field == null) {
+            failureCollector.addFailure(String.format("The %s field '%s' does not exist in input schema.",
+                                                      IS_TARGET_FIELD, isTargetField),
+                                        null).withConfigElement(IS_TARGET_FIELD, isTargetField);
+          } else {
+            Schema schema = field.getSchema();
+            if (!Schema.Type.BOOLEAN.equals(schema.getType())) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Plugin.java]: if the field is a nullable boolean, this leads to a confusing message complaining that the field is not a boolean.

It would be better to explicitly check for a nullable boolean and say that it must be non-null. The alternative is to treat null as false.

In either case, still need to check for nullables.

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 4]
```diff
@@ -89,6 +95,69 @@ protected StructuredRecord computeNext() {
...
     }
 
+    // get the next record that is available to compute the end time, it must be a record with same id
+    // that will not get dedup, or null if cur record is last record/
+    // This is needed because of the following scenarios:
+    // 1(cur): id: 1, name: john, isTarget: true (should be preserved)
+    // 2: id: 1, name: john, isTarget: false (should get dedup)
+    // 3: id: 1, name: sam, isTarget: false (should get dedup)
+    // 4: id: 1, name: sam, isTarget: false (should be preserved)
+    // here first and fourth record will be preserved, so when we compute the end date of 1st record, we should use
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: I'm not really following the logic here. Why do we need nextNext and why do we need 2 dedup loops? Why isn't the change in the loop above enough? Shouldn't cur be set to record 1 and prev be set to record 4 at that point?

**Review by albertshau — COMMENTED** [review]
also be sure to update the docs

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 4 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: No, prev should always be before the cur record. 
The first loop is to move the iterator to find the correct cur, the second loop is to move the iterator to find the correct next.

If you consider the the following example:
1: id: 1, name: john, isTarget: false, startTime: 0 (should get dedup)
2: id: 1, name: john, isTarget: true, startTime: 10(should be preserved)
3: id: 1, name: john, isTarget: false, startTime: 100 (should get dedup)
4: id: 1, name: sam, isTarget: false, startTime:1000 (should get dedup)
5: id: 1, name: sam, isTarget: false, startTime:10000 (should be preserved)

After the first loop, 2# will be found as cur record, at this time, prev points to 1#, and next points to 3#. But 3# is a record should get discarded because it is same as 2#. So a second loop is needed to check a next record which either a record which is in target table, or a record that should not be discarded, which means it is must be different from the cur record(2#), but also different from the following records.(5#) Thats why we need to have nextNext and a second loop. 

After deduplicating and calculating the end time, above records should generate:
2: id: 1, name: john, isTarget: true, startTime: 10, endTime: 9999(10000-1)
5: id: 1, name: sam, isTarget: false, startTime:10000, endTime: Active

I tried to put the code in one loop, and the code will be super hard to follow, since that loop involves updating prev, cur, next, nextNext at the same time as well as comparing the prev and nextNext. That's why I break down this to a second loop to make logic easier to follow 

We cannot simply use the start time of the next record after first loop to compute the end time because of the following scenario:
1: id: 1, name: john, isTarget: false, startTime: 0 (should get dedup)
2: id: 1, name: john, isTarget: true, startTime: 10(should be preserved)
3: id: 1, name: john, isTarget: false, startTime: 100 (should get dedup)
4: id: 1, name: john, isTarget: false, startTime: 1000 (should get dedup)

For 2#, if we simply use 3# to calculate the result, then after deduplicate, we will get:
2: id: 1, name: john, isTarget: true, startTime: 10, endTime: 99

There is no active record since both 3# and 4# will get deduplicated.

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 1 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: A lot of lines are just comments, don't think break it down makes things easier to read

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 4 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: @albertshau I updated the doc as well as updating this comment about the behavior, we can discuss about it if you want.

File: docs/SCD2-sparkcompute.md [hunk 5]
```diff
@@ -39,7 +39,9 @@ After filling in null, the resulting records will be:
 | 1  | 100        | John | abc@abc.com   |
 
 **Deduplicate:** Deduplicate consecutive records that have no changes. This can be used with the blacklist to exclude some fields from comparison.
-The blacklist typically includes the start time field. For example, suppose the following records are processed with start_date as blacklist:
+The blacklist typically includes the start time field. If Perserve Target is set to false, the record with the latest record will be picked.
+The end time of a record will be computed as the start time of the next non-deduplicated record minus the offset.
```
Comment (albertshau) [inline docs/SCD2-sparkcompute.md:43]: end time is mentioned here, but is not in the example below. Can you add it so that it is more clear

File: docs/SCD2-sparkcompute.md [hunk 6]
```diff
@@ -56,6 +58,26 @@ After deduplicating, the resulting records will be:
 
 **Blacklist:** Blacklist for fields to ignore to compare when deduplicating the record.
 
+**Preserve Target** Preserve all the records from the scd2 target table when deduplicating.
+
+**Target Field** The name of the field that tells whether the record is coming from the scd2 target table. This field must be boolean.
+For example, suppose the following records are processed with Deduplicate and Preserve Target set to true, with Blacklist set to start_date, is_target:
```
Comment (albertshau) [inline docs/SCD2-sparkcompute.md]: If the target field is given, I think it's fine to assume (from a documentation standpoint) that dedupe and preserve are set to true. I think it would be more clear to say:

For example, suppose the target field is set to 'is_target' and the following records are processed:

File: docs/SCD2-sparkcompute.md [hunk 7]
```diff
@@ -56,6 +58,26 @@ After deduplicating, the resulting records will be:
 
 **Blacklist:** Blacklist for fields to ignore to compare when deduplicating the record.
 
+**Preserve Target** Preserve all the records from the scd2 target table when deduplicating.
+
+**Target Field** The name of the field that tells whether the record is coming from the scd2 target table. This field must be boolean.
```
Comment (albertshau) [inline docs/SCD2-sparkcompute.md:63]: This is only used if preserve target is set to true. Can you add a filter to the widget so this is hidden unless preserve = true?

See https://github.com/data-integrations/http/blob/develop/widgets/HTTP-batchsource.json#L501 for some examples

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 4 above)
Comment (albertshau) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: I'm still having a hard time following this logic. It seems difficult to understand without actually stepping through examples. I think it would be cleaner if we did a few things:

1.  Wrap the Iterator given in the constructor with a PeekingIterator (from guava). Then you can just peek() instead of maintaining extra state yourself.

2. Refactor the dedupe check into a separate method

3. Use descriptive variable names instead of calling ._1() and ._2() inline.

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 8]
```diff
@@ -60,22 +62,64 @@ protected StructuredRecord computeNext() {
...
+        Object value = curRecord.get(fieldName);
         if (blacklist.contains(fieldName)) {
           continue;
         }
 
         // check if there is difference between next record and cur record
-        Object nextVal = next._2().get(fieldName);
+        StructuredRecord nextRecord = next._2();
+        Object nextVal = nextRecord.get(fieldName);
         if ((nextVal == null) != (value == null) || (value != null && !value.equals(nextVal))) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: Objects.equals(value, nextValue)

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 9]
```diff
@@ -60,22 +62,64 @@ protected StructuredRecord computeNext() {
...
+      }
+
+      SCD2Key curKey = cur._1();
+      SCD2Key nextKey = next._1();
+      if (next == null || !nextKey.equals(curKey)) {
         break;
       }
 
-      for (Schema.Field field : cur._2().getSchema().getFields()) {
+      for (Schema.Field field : curRecord.getSchema().getFields()) {
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: refactor this into a separate recordsEqual() method for checking record equality, ignoring fields in the blacklist

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 10]
```diff
@@ -60,22 +62,64 @@ protected StructuredRecord computeNext() {
...
+   * 1. id: 1, name: john, isTarget: false (should get dedup)
+   * 2. id: 1, name: john, isTarget: true (should be preserved)
+   * 3. id: 1, name: john, isTarget: false (should get dedup)
+   * 4. id: 1, name: sam, isTarget: false (should get dedup)
+   * 5. id: 1, name: sam, isTarget: false (should be preserved)
+   *
+   * After this method, the prev, cur, next will be pointed to:
+   * 1.(prev) id: 1, name: john, isTarget: false (should get dedup)
+   * 2.(cur) id: 1, name: john, isTarget: true (should be preserved and returned)
+   * 3.(next) id: 1, name: john, isTarget: false (should get dedup)
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java]: I think it's better if the contract is that the iterator advances to 4 after the method is called and not 3. Then the code becomes something like:

```
private Tuple2<SCD2Key, StructuredRecord> dedupeNext() {
  Tuple2<SCD2Key, StructuredRecord> currentElement = records.next();

  while (conf.deduplicate() && records.hasNext()) {
    Tuple2<SCD2Key, StructuredRecord> nextElement = records.peek();
    if (isTarget(nextElement)) {
      currentElement = records.next();
      continue;
    }
    if (isDiff(currentElement, nextElement) {
      break;
    }
    records.next();
  }

  return currentElement;
}

```

Then you don't need need the other method with its prev and next tracking, and don't need to modify state anywhere except in the computeNext() method. You also only need to call this once per computeNext() call (except for the very first time, when you need to call it once to initialize cur and once to initialize next).

```
protected StructuredRecord computeNext() {
  ...
  prev = cur;
  if (prev = null) {
    cur = dedupeNext();
  } else {
    cur = next;
  }
  next = records.hasNext() ? dedupeNext() : null;
}

```

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java [hunk 11]
```diff
@@ -57,50 +60,92 @@ protected StructuredRecord computeNext() {
...
+    Tuple2<SCD2Key, StructuredRecord> result = currentElement;
+    boolean findResult = false;
+
+    while (conf.deduplicate() && records.hasNext()) {
+      Tuple2<SCD2Key, StructuredRecord> nextElement = records.peek();
+      if (isTarget(currentElement._2())) {
+        findResult = true;
+      }
+
+      // if the current is different from next or if the next is from target table, we can stop deduplicating
```
Comment (albertshau) [inline src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java:110]: next is not expected to be a target right? This is just in case there happen to be dupes in the target?

**Review by albertshau — APPROVED** [review]
lgtm

File: src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java (hunk 11 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java:110]: Yes, this is just to guard against in case the target contains some duplicates. A unit test was also added for this.
